cloud/google/compute/reconciler.go

```diff
@@ -0,0 +1,230 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package compute
+
+import (
+	"os"
+	"os/signal"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/kris-nova/kubicorn/apis/cluster"
+	"github.com/kris-nova/kubicorn/cloud"
+	"github.com/kris-nova/kubicorn/cloud/google/compute/googleSDK"
+	"github.com/kris-nova/kubicorn/cloud/google/compute/resources"
+	"github.com/kris-nova/kubicorn/cutil/hang"
+	"github.com/kris-nova/kubicorn/cutil/logger"
+)
+
+var sigCaught = false
+
+// Todo add description.
+type Reconciler struct {
+	Known *cluster.Cluster
+}
+
+// NewReconciler creates a new Reconciler using the expected cluster.
+func NewReconciler(expected *cluster.Cluster) cloud.Reconciler {
+	return &Reconciler{
+		Known: expected,
+	}
+}
+
+var model map[int]cloud.Resource
+
+// Init is used to create the sdk and add this to the resources.
+func (r *Reconciler) Init() error {
+	sdk, err := googleSDK.NewSdk()
+	if err != nil {
+		return err
+	}
+
+	resources.Sdk = sdk
+	model = ClusterModel(r.Known)
+	return nil
+}
+
+// GetActual is used to create a representation of the cluster existing on the cloud provider.
+func (r *Reconciler) GetActual() (*cluster.Cluster, error) {
+	actualCluster := newClusterDefaults(r.Known)
+	for i := 0; i < len(model); i++ {
+		resource := model[i]
+		actualResource, err := resource.Actual(r.Known)
+		if err != nil {
+			return nil, err
+		}
+		actualCluster, err = resource.Render(actualResource, actualCluster)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return actualCluster, nil
+}
+
+// GetExpected is used to create a representation of the cluster expected on the cloud provider.
+func (r *Reconciler) GetExpected() (*cluster.Cluster, error) {
+	expectedCluster := newClusterDefaults(r.Known)
+	for i := 0; i < len(model); i++ {
+		resource := model[i]
+		expectedResource, err := resource.Expected(r.Known)
+		if err != nil {
+			return nil, err
+		}
+		expectedCluster, err = resource.Render(expectedResource, expectedCluster)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return expectedCluster, nil
+}
+
+func cleanUp(cluster *cluster.Cluster, i int) error {
+	logger.Warning("--------------------------------------")
+	logger.Warning("Attempting to delete created resources!")
+	logger.Warning("--------------------------------------")
+	for j := i - 1; j >= 0; j-- {
+		var err error
+		resource := model[j]
+		createdResource := createdResources[j]
+		_, err = resource.Delete(createdResource, cluster)
+		if err != nil {
+			j, err = destroyI(err, j)
+			if err != nil {
+				return err
+			}
+			continue
+		}
+	}
+	return nil
+}
+
+var createdResources = make(map[int]cloud.Resource)
```

Comment:
Would it make more sense to have this be a property of the reconciler rather than a singleton?